federated_testing.md

```diff
@@ -0,0 +1,96 @@
+
+The quality of Kubernetes depends on exercising the software on a
+variety of platforms.  The Kubernetes project welcomes contributions of
+test results from organizations that execute e2e test jobs.  The
+[Kubernetes test history
+dashboard](http://storage.googleapis.com/kubernetes-test-history/static/index.html)
+displays the latest results from all federated test jobs.
+
+
+## Overview
+
+To contribute test results for a platform, an organization executes the
+e2e test cases and uploads the test results to a Google Cloud Storage
+(GCS) bucket.  Kubernetes project contains scripts in `hack/` to run the
+tests and to upload the results.
+
+The test history dashboard is generated periodically by scripts in this
+repository that read test results from all of the GCS buckets.  The test
+history shows test results from jobs run during the previous 24 hours.
+A test job should execute at least once every 24 hours.
+
+The remainder of this document explains how to contribute test results.
+
+
+### Run the e2e tests and upload results
+
+Setup a periodic test job that will run at least once per day.  The test
+job runs e2e tests and uploads the results.
+
+1. Create a Google Cloud Storage bucket that will store the test
+   results.  The bucket should have public read access.  The GCS URL
+   must be stored in an environment variable named
+   JENKINS_GCS_LOGS_PATH.  For example, Google's Jenkins server stores
+   results to `gs://kubernetes-jenkins/logs/`.
+2. Run `hack/jenkins/e2e-runner.sh`.  There are several environment
+   variables that affect the behavior of this script.  See [e2e-runner
+   Environment Variables](#e2e-runner-environment-variables).
+3. Run `JENKINS_BUILD_FINISHED={SUCCESS|UNSTABLE|FAILURE|ABORTED}
+   hack/jenkins/upload-to-gcs.sh` to upload the results to the GCS
+   bucket.  If running from a Jenkins job, it is recommend to perform
+   this step as a post-build step.  See
+   `jenkins/job-configs/global.yaml` for an example.
+
+
+### Include results in test history
+
+Modify scripts that collect results from federate test jobs to include
+the Google Cloud Storage bucket in the list of buckets that are read.
+
+1. Add the name of the Google Cloud Storage bucket to
+   `BUCKET_WHITELIST` in [gubernator/main.py](gubnerator/main.py).
+2. Add an entry for the test job in
+   [`jenkins/test-history/buckets.json`](jenkins/test-history/buckets.json).
+   The key for each entry is a Google Cloud Storage bucket URL.  Each
+   bucket has a `prefix` field that used to identify tests in the
+   test history dashboard.
+
+
+### Results bucket layout
+
+The `upload-to-gcs.sh` script stores results in a specified GCS bucket.
+Each bucket has a sub-directory for each job (`${JOB_NAME}`) that
+contains a `latest-build.txt` file and a directory with results for each
+build of the test job.  The results sub-directories are numeric values
+named for the value of `${BUILD_NUMBER}` environment variable that is
+set when the e2e tests are run.
+
```

Comment:
I think that the documentation for upload-to-gcs.sh should remain in the repository that contains the script.  I wasn't aware of the other document.  Maybe I can just remove this section, rather than duplicate it.